python/lsst/daf/butler/registry/queries/exprParser/parserYacc.py

```diff
@@ -27,19 +27,107 @@
 # -------------------------------
 #  Imports of standard modules --
 # -------------------------------
+import re
 
 # -----------------------------
 #  Imports for other modules --
 # -----------------------------
-from .exprTree import (BinaryOp, Identifier, IsIn, NumericLiteral,
-                       Parens, StringLiteral, RangeLiteral, UnaryOp)
+import astropy.time
+from .exprTree import (BinaryOp, Identifier, IsIn, NumericLiteral, Parens,
+                       RangeLiteral, StringLiteral, TimeLiteral, UnaryOp)
 from .ply import yacc
 from .parserLex import ParserLex
 
 # ----------------------------------
 #  Local non-exported definitions --
 # ----------------------------------
 
+# The purpose of this regex is to guess time format if it is not explicitly
+# provided in the string itself
+_re_time_str = re.compile(r"""
+    ((?P<format>\w+)/)?             # optionally prefixed by "format/"
+    (?P<value>
+        (?P<number>-?(\d+(\.\d*)|(\.\d+)))   # floating point number
+        |
+        (?P<iso>\d+-\d+-\d+([ T]\d+:\d+(:\d+([.]\d*)?)?)?)   # iso(t)
+        |
+        (?P<fits>[+]\d+-\d+-\d+(T\d+:\d+:\d+([.]\d*)?)?)   # fits
+        |
+        (?P<yday>\d+:\d+(:\d+:\d+(:\d+([.]\d*)?)?)?)         # yday
+    )
+    (/(?P<scale>\w+))?              # optionally followed by "/scale"
+    $
+""", re.VERBOSE | re.IGNORECASE)
+
+
+def _parseTimeString(time_str):
+    """Try to convert time string into astropy.Time.
+
+    Parameters
+    ----------
+    time_str : `str`
+        Input string.
+
+    Returns
+    -------
+    time : `astropy.time.Time`
+
+    Raises
+    ------
+    ValueError
+        Raised if input string has unexpected format
+    """
+    match = _re_time_str.match(time_str)
+    if not match:
+        raise ValueError(f"Time string \"{time_str}\" does not match known formats")
+
+    value, fmt, scale = match.group("value", "format", "scale")
+    if fmt is not None:
+        fmt = fmt.lower()
+        if fmt not in astropy.time.Time.FORMATS:
+            raise ValueError(f"Time string \"{time_str}\" specifies unknown time format \"{fmt}\"")
```

Comment:
We raise early rather than waiting for astropy because we don't like the astropy exception?

Reply:
I do it to make more specific message about format, guessing that format is wrong from astropy exception would be difficult.

Reply:
Interesting. Bad format:
```
>>> t.Time(123456.43, format="blah")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/time/core.py", line 429, in __init__
    precision, in_subfmt, out_subfmt)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/time/core.py", line 484, in _init_from_vals
    precision, in_subfmt, out_subfmt)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/time/core.py", line 526, in _get_time_fmt
    sorted(self.FORMATS)))
ValueError: Format 'blah' is not one of the allowed formats ['byear', 'byear_str', 'cxcsec', 'datetime', 'datetime64', 'decimalyear', 'fits', 'gps', 'iso', 'isot', 'jd', 'jyear', 'jyear_str', 'mjd', 'plot_date', 'stardate', 'unix', 'yday', 'ymdhms']
```
but bad scale:
```
>>> t.Time(123456.43, scale="xc")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/time/core.py", line 429, in __init__
    precision, in_subfmt, out_subfmt)
  File "/Users/timj/work/lsstsw3/miniconda/envs/lsst-scipipe/lib/python3.7/site-packages/astropy/time/core.py", line 476, in _init_from_vals
    sorted(self.SCALES)))
astropy.time.core.ScaleValueError: Scale 'xc' is not in the allowed scales ['local', 'tai', 'tcb', 'tcg', 'tdb', 'tt', 'ut1', 'utc']
```
So they have a special exception for bad time scale but not for bad format.